zkwrapper/source/zkwrapper.cc

```diff
@@ -501,3 +542,37 @@ std::vector <uint8_t> ZKWrapper::get_byte_vector(const std::string &string) {
 void ZKWrapper::close() {
 	zookeeper_close(zh);
 }
+
+bool ZKWrapper::flush(const std::string& full_path, bool synchronous) const {
+	// flush is only blocking when the synchronous flag is set
+	if (synchronous) {
+
+		// I tried using condition variables, but it ended up failing on occasion, so sadly I resort to polling :(
+		bool * flag = (bool *)malloc(sizeof(bool));
+
+		// Lambda to call on function completion
+		string_completion_t completion = [](int rc, const char *value, const void *data){
+			bool * flag_ptr = const_cast<bool *>(reinterpret_cast<const bool *>(data));
+			* flag_ptr = true;
+			std::this_thread::sleep_for (std::chrono::milliseconds(1000));
```

Comment:
This is a minimal effort at preventing the memory leak. I sleep before delete in hope that the blocked function awakes and exits before we attempt to free this flag. This is very ugly, but basically I am not sure if there are any super good ways to guarantee that we do not free this variable before the blocking function attempts to read it